apps/src/labs/projects/ProjectManager.ts

```diff
@@ -0,0 +1,163 @@
+/**
+ * ProjectManager manages loading and saving projects that have a source
+ * and channel component, and are indexed on a channel id (or other unique identifier).
+ * It accepts a sources and channels store, which handle communication with the relevant
+ * apis for loading and saving sources and channels.
+ *
+ * ProjectManager throttles saves to only occur once every 30 seconds, unless a force
+ * save is requested. If save is called within 30 seconds of the previous save,
+ * the save is queued and will be executed after the 30 second interval has passed.
+ */
+import {SourcesStore} from './SourcesStore';
+import {ChannelsStore} from './ChannelsStore';
+import {Project} from '../types';
+import {AppOptionsStore} from '../AppOptionsStore';
+
+export enum ProjectManagerEvent {
+  SaveStart,
+  SaveNoop,
+  SaveSuccess,
+  SaveFail,
+}
+
+export default class ProjectManager {
+  channelId: string;
+  sourcesStore: SourcesStore;
+  channelsStore: ChannelsStore;
+  appOptionsStore: AppOptionsStore;
+  getProject: () => Project;
+
+  private nextSaveTime: number | null = null;
+  private readonly saveInterval: number = 30 * 1000; // 30 seconds
+  private saveInProgress = false;
+  private saveQueued = false;
+  private eventListeners: {
+    [key in keyof typeof ProjectManagerEvent]?: [(response: Response) => void];
+  } = {};
+
+  constructor(
+    channelId: string,
+    sourcesStore: SourcesStore,
+    channelsStore: ChannelsStore,
+    appOptionsStore: AppOptionsStore,
+    getProject: () => Project
+  ) {
+    this.channelId = channelId;
+    this.sourcesStore = sourcesStore;
+    this.channelsStore = channelsStore;
+    this.appOptionsStore = appOptionsStore;
+    this.getProject = getProject;
+  }
+
+  async load(): Promise<Response> {
+    const sourceResponse = await this.sourcesStore.load(this.channelId);
+    // If sourceResponse is not ok, we still want to load the channel. Source can
+    // return not found if the project is new.
+    let source = {};
+    if (sourceResponse.ok) {
+      source = await sourceResponse.json();
+    }
+
+    const channelResponse = await this.channelsStore.load(this.channelId);
+    if (!channelResponse.ok) {
+      return channelResponse;
+    }
+
+    const channel = await channelResponse.json();
+    // ensure the project type is set on the channel
+    channel.projectType = this.appOptionsStore.getProjectType();
+    const project = {source, channel};
+    const blob = new Blob([JSON.stringify(project)], {
+      type: 'application/json',
+    });
+    return new Response(blob);
+  }
+
+  hasQueuedSave(): boolean {
+    return this.saveQueued;
+  }
+
+  // TODO: Add functionality to reduce channel updates during
```

Comment:
It looks like there is some interesting (and important) logic in which we first attempt to save sources, and only if that succeeds do we then attempt to save the channel.  Could we capture the thinking behind this behavior in a comment for future readers?